main/src/main/scala/sbt/MainLoop.scala

```diff
@@ -155,13 +155,72 @@ object MainLoop {
         state.log error errMsg
         state.fail
     }
-    StandardMain.exchange publishEventMessage ExecStatusEvent(
+    val exitCode = determineExitCode(state, nextState)
+    val doneEvent = ExecStatusEvent(
       "Done",
       channelName,
       exec.execId,
-      newState.remainingCommands.toVector map (_.commandLine))
+      newState.remainingCommands.toVector map (_.commandLine),
+      exitCode,
+    )
+    if (doneEvent.execId.isDefined) { // send back a response or error
+      import sbt.protocol.codec.JsonProtocol._
+      StandardMain.exchange publishEvent doneEvent
+    } else { // send back a notification
+      StandardMain.exchange publishEventMessage doneEvent
+    }
     newState
   }
 
   def logFullException(e: Throwable, log: Logger): Unit = State.logFullException(e, log)
-}
+
+  private def determineExitCode(state: State, newState: State): Option[Int] = {
+    sealed trait ExitCode
```

Comment:
See https://github.com/sbt/sbt/blob/1.x/protocol/src/main/scala/sbt/internal/langserver/ErrorCodes.scala

Reply:
You're currently making up 0 or 1 based numbering that gets sent to LSP client correct?
http://www.jsonrpc.org/specification determines the number you have to use.

Reply:
it says:

> The remainder of the space is available for application defined errors.

I'm using 1 when there is some unknown error that was handled by onFailure, that's it.

Reply:
VS Code uses -32800. Maybe it's better to extend to some large negative integers following the convention, like `-33000`.

Reply:
why is `-33000` better than `1` for "unknown error"?

Reply:
From the point of view of the LSP client (let's say Vim or something) they need to handle various error events. I want to fit into the ecosystem harmoniously.